### ISSUE TITLE
current-when not working with dynamic segment

### ISSUE THREAD

**Issue by amk221** [emberjs/ember.js#14615]
Here is an example:

https://ember-twiddle.com/93563986146cde7c19cb7e9e985f32c0?openFiles=templates.application.hbs%2C&route=%2Ffoo%2F1

**Comment by locks** [comment]
@amk221 I don't see any problem, can you please describe your expectation and what's actually happening?

**Comment by amk221** [comment]
Sure, the hyperlink to `foo` has `current-when='foo.index foo.bar'`
Therefore, I am expecting that hyperlink to be _bold_ (active) when I visit `foo.bar`
But it does not.

Remove the dynamic segment from the link-to, and it works as expected.

**Comment by locks** [comment]
Ah, it's `current-when=['foo.index', 'foo.bar']`. Not that I recommend it :P

**Comment by amk221** [comment]
Why do you not recommend it?
Why is it not documented?
Why does it work when removing the dynamic segment?

**Comment by locks** [comment]
> Why do you not recommend it?
> Why is it not documented?
Because it should ideally be done via route nesting/architecture.

> Why does it work when removing the dynamic segment?
I don't know what you mean

**Comment by amk221** [comment]
> Because it should ideally be done via route nesting/architecture.

If I am doing something ill-advised, can you link me to how I should be doing it?

> Why does it work when removing the dynamic segment?
> I don't know what you mean
I mean, if you remove the number "1" from the `link-to` it will highlight correctly.

**Comment by locks** [comment]
Oh, you were passing two arguments to a route (`foo`) that only has one dynamic segment.

**Comment by amk221** [comment]
:) nope, I'm not the arguments are correct. 
'foo' 'is the title text
'foo' is the route
'1' is the dynamic arg

**Comment by psteininger** [comment]
@amk221 did you ever get around this issue? I am running into the same situation, on a link with 2 dynamic elements `/facilities/:facility_id/issues/:issue_id` on `/facilities/:facility_id/issues` it works, but on `/facilities/:facility_id/issues/:issue_id` it doesn't despite:
```hbs
{{#link-to 'facility.issues' model current-when='facility.issues facility.issue'}}
```
the routes are like so:
```js
...
 this.route('facilities', {resetNamespace: true});
    this.route('facility', {resetNamespace: true, path: '/facilities/:facility_id'}, function () {
      ...
      this.route('issues');
      this.route('issue', {path: '/issues/:issue_id'});
    });
...
```
ignore the `resetNamespace:true` it deals with being under a `protected` route

**Comment by amk221** [comment]
Nope, sorry I managed to write a failing test, but I have no clue as to how and fix it

**Comment by psteininger** [comment]
@amk221 It looks like the problem sits in the fact that for `facility.issue` route there need to be 2 `models`/`contexts`, but in the link definition there is only 1. I did some debugging and saw that for another `link-to` (which has both the facility and the issue) the method would have succeeded.

I traced it down to this call: 
```js
// file: 'ember-routing/system/router_state' #isActiveIntent
 if (!this.routerJs.isActiveIntent(routeName, models, null, state)) {
        return false;
      }
```
that returns false

**Comment by psteininger** [comment]
I think I traced it down to this method: 
```js
function handlerInfosEqual(handlerInfos, otherHandlerInfos) {
  if (handlerInfos.length !== otherHandlerInfos.length) {
    return false;
  }

  for (var i = 0, len = handlerInfos.length; i < len; ++i) {
    if (handlerInfos[i] !== otherHandlerInfos[i]) {
      return false;
    }
  }
  return true;
}
```
The key difference is that there is a `_names` key with array of params (in this case `issue_id`)

**Comment by pixelhandler** [comment]
@amk221 @locks @mmun @psteininger is this still an issue, perhaps we should close or create a new reproduction of this, what do you think?

**Comment by amk221** [comment]
FWIW the twiddle demonstrates it's still an issue, so does the failing test I wrote, as mentioned previously I don't have enough knowledge of the area to fix it.

**Comment by psteininger** [comment]
@pixelhandler @amk221 I missed earlier messages (Gmail sticks them deep in these silly tabs). This is still an issue. I don't remember the nitty gritty of the issue, but the above `handlerInfosEqual` was at the root of the issue.

**Comment by gokatz** [comment]
This issue is still reproducible: https://ember-twiddle.com/cd19e5c2abd8a14ea56fba2754d79227?openFiles=templates.application.hbs%2C

Any help on this?
